In [12]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/cleaned-token/vocab.txt
/kaggle/input/generated-model/model.h5
/kaggle/input/movie-review/movie_reviews-meta.json
/kaggle/input/movie-review/README
/kaggle/input/movie-review/movie_review.csv
/kaggle/input/movie-review/movie_reviews/movie_reviews/README
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv537_13516.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv062_24556.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv735_20218.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv440_16891.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv221_27081.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv205_9676.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv857_17527.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv689_13701.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg/cv935_24977.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/neg

/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv808_12635.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv305_9946.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv864_3416.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv714_18502.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv386_10080.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv460_10842.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv970_18450.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv260_13959.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv349_13507.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv112_11193.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv980_10953.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv031_18452.txt
/kaggle/input/movie-review/movie_reviews/movie_reviews/pos/cv642_29867.txt
/kaggle/input/movie-review/

# Tutorial Overview

This tutorial is divided into the following parts:
1. Movie Review Dataset
2. Data Preparation
3. Train CNN With Embedding Layer 
4. Evaluate Model

## 1. Movie Review Dataset

In this tutorial, I will use the Movie Review Dataset. The Movie Review Data is a collection of movie reviews retrieved from the imdb.com website in the early 2000s by Bo Pang and Lillian Lee. The reviews were collected and made available as part of their research on natural language processing. The reviews were originally released in 2002, but an updated and cleaned up version was released in 2004, referred to as v2.0. The dataset is comprised of 1,000 positive and 1,000 negative movie reviews drawn from an archive of the rec.arts.movies.reviews newsgroup hosted at IMDB. The authors refer to this dataset as the polarity dataset.

## Load dataset

In [13]:
# open the file as read only file
filename = '../input/movie-review/movie_reviews/movie_reviews/neg/cv000_29416.txt'
file = open(filename, 'r')
text = file.read()
# file.close()

The above snippet loads the texts as an ASCII document where any white spaces are still there. So, we can refactor that into a more generic function to load any document that would help in future imports. 

In [14]:
def load_doc(filename):
    file = open(filename, 'r')
    text = file.read()
#     file.close()
    return text

We have two directories, neg and pos each of which has a text files that need to be read. So, the function `load_doc()` will be responsible for doing that for us. We will getting each file from each directory through the `listdir()`. I will do it in another function to make it more generic.

In [15]:
from os import listdir

def load_dir(directory):
    for filename in listdir(directory):
        #skip files that doesn't have the right extension
        if not filename.endswith('.txt'):
            next
        # create the full path of the file to open
        path = directory + '/' + filename
        # load document
        doc = load_doc(path)
        print('loading %s' % filename)
        
directory = '../input/movie-review/movie_reviews/movie_reviews/neg'
load_dir(directory)

loading cv537_13516.txt
loading cv062_24556.txt
loading cv735_20218.txt
loading cv440_16891.txt
loading cv221_27081.txt
loading cv205_9676.txt
loading cv857_17527.txt
loading cv689_13701.txt
loading cv935_24977.txt
loading cv149_17084.txt
loading cv226_26692.txt
loading cv644_18551.txt
loading cv742_8279.txt
loading cv256_16529.txt
loading cv014_15600.txt
loading cv839_22807.txt
loading cv171_15164.txt
loading cv580_15681.txt
loading cv232_16768.txt
loading cv583_29465.txt
loading cv317_25111.txt
loading cv659_21483.txt
loading cv920_29423.txt
loading cv837_27232.txt
loading cv164_23451.txt
loading cv953_7078.txt
loading cv830_5778.txt
loading cv203_19052.txt
loading cv733_9891.txt
loading cv401_13758.txt
loading cv592_23391.txt
loading cv682_17947.txt
loading cv629_16604.txt
loading cv158_10914.txt
loading cv370_5338.txt
loading cv126_28821.txt
loading cv614_11320.txt
loading cv445_26683.txt
loading cv478_15921.txt
loading cv788_26409.txt
loading cv095_28730.txt
loading cv558_29376.tx

loading cv321_14191.txt
loading cv065_16909.txt
loading cv452_5179.txt
loading cv982_22209.txt
loading cv848_10061.txt
loading cv740_13643.txt
loading cv267_16618.txt
loading cv747_18189.txt
loading cv730_10729.txt
loading cv151_17231.txt
loading cv966_28671.txt
loading cv168_7435.txt
loading cv690_5425.txt
loading cv589_12853.txt
loading cv771_28466.txt
loading cv697_12106.txt
loading cv189_24248.txt
loading cv002_17424.txt
loading cv782_21078.txt
loading cv105_19135.txt
loading cv894_22140.txt
loading cv372_6654.txt
loading cv990_12443.txt
loading cv017_23487.txt
loading cv746_10471.txt
loading cv781_5358.txt
loading cv643_29282.txt
loading cv053_23117.txt
loading cv233_17614.txt
loading cv588_14467.txt
loading cv467_26610.txt
loading cv354_8573.txt
loading cv045_25077.txt
loading cv731_3968.txt
loading cv028_26964.txt
loading cv192_16079.txt
loading cv023_13847.txt
loading cv822_21545.txt
loading cv195_16146.txt
loading cv116_28734.txt
loading cv393_29234.txt
loading cv209_28973.txt

loading cv238_14285.txt
loading cv458_9000.txt
loading cv077_23172.txt
loading cv979_2029.txt
loading cv525_17930.txt
loading cv819_9567.txt
loading cv409_29625.txt
loading cv616_29187.txt
loading cv125_9636.txt
loading cv805_21128.txt
loading cv615_15734.txt
loading cv147_22625.txt
loading cv482_11233.txt
loading cv311_17708.txt
loading cv446_12209.txt
loading cv732_13092.txt
loading cv512_17618.txt
loading cv755_24881.txt
loading cv564_12011.txt
loading cv144_5010.txt
loading cv015_29356.txt
loading cv100_12406.txt
loading cv032_23718.txt
loading cv036_18385.txt
loading cv623_16988.txt
loading cv271_15364.txt
loading cv939_11247.txt
loading cv656_25395.txt
loading cv796_17243.txt
loading cv880_29629.txt
loading cv593_11931.txt
loading cv225_29083.txt
loading cv056_14663.txt
loading cv135_12506.txt
loading cv030_22893.txt
loading cv869_24782.txt
loading cv722_7571.txt
loading cv758_9740.txt
loading cv081_18241.txt
loading cv892_18788.txt
loading cv543_5107.txt
loading cv397_28890.txt


## Clean Dataset

In this section, we will look at what data cleaning we might want to do to the movie review data. We will assume that we will be using a bag-of-words model or perhaps a word embedding that does not require too much preparation.

The first step I would like to take is to look at a sample of files to see what sort of words included in the text file. Doing so, would make it more easier to plan ahead which cleaning approach should be taken. I will use the function I did before `load_doc()` then apply the `split()` function to it. 

In [16]:
filename = '../input/movie-review/movie_reviews/movie_reviews/neg/cv000_29416.txt'
tokens = load_doc(filename).split()
print(tokens)

['plot', ':', 'two', 'teen', 'couples', 'go', 'to', 'a', 'church', 'party', ',', 'drink', 'and', 'then', 'drive', '.', 'they', 'get', 'into', 'an', 'accident', '.', 'one', 'of', 'the', 'guys', 'dies', ',', 'but', 'his', 'girlfriend', 'continues', 'to', 'see', 'him', 'in', 'her', 'life', ',', 'and', 'has', 'nightmares', '.', "what's", 'the', 'deal', '?', 'watch', 'the', 'movie', 'and', '"', 'sorta', '"', 'find', 'out', '.', '.', '.', 'critique', ':', 'a', 'mind-fuck', 'movie', 'for', 'the', 'teen', 'generation', 'that', 'touches', 'on', 'a', 'very', 'cool', 'idea', ',', 'but', 'presents', 'it', 'in', 'a', 'very', 'bad', 'package', '.', 'which', 'is', 'what', 'makes', 'this', 'review', 'an', 'even', 'harder', 'one', 'to', 'write', ',', 'since', 'i', 'generally', 'applaud', 'films', 'which', 'attempt', 'to', 'break', 'the', 'mold', ',', 'mess', 'with', 'your', 'head', 'and', 'such', '(', 'lost', 'highway', '&', 'memento', ')', ',', 'but', 'there', 'are', 'good', 'and', 'bad', 'ways', 'of'

From the results, I can see that I need to remove 
1. Punctuations, 
2. numbers, 
3. One word character
4. English stopping words
5. Punctuation words 

I will put those all together in a function called `clean_doc()`

In [17]:
from nltk.corpus import stopwords
import string
import re

def clean_doc(doc):
    """
    input: file name in a text format
    what it does: clean up the passing document from punctuation, stop words,...etc
    output: cleaned document saved in a text file
    """
    #split the given document into tokens 
    tokens = doc.split()
    # prepare the the regex to filter out the punctuation from tokens
    punc_filter = re.compile('%s' %re.escape(string.punctuation))
    # remove punctuation from each word
    tokens = [punc_filter.sub('', w) for w in tokens]
    # remove remaining tokens that are not alphabetic
    tokens = [word for word in tokens if word.isalpha()]
    # filter out stop words
    stop_words = set(stopwords.words('english'))
    tokens = [w for w in tokens if not w in stop_words]
    # filter out short tokens
    tokens = [word for word in tokens if len(word) > 1]
    return tokens

In [18]:
filename = '../input/movie-review/movie_reviews/movie_reviews/neg/cv000_29416.txt'
text = load_doc(filename)
tokens = clean_doc(text)
print(tokens[:20])

['plot', 'two', 'teen', 'couples', 'go', 'church', 'party', 'drink', 'drive', 'get', 'accident', 'one', 'guys', 'dies', 'girlfriend', 'continues', 'see', 'life', 'nightmares', 'deal']


## Define a vocabulary

It is important to define a vocabulary of known words when using a text model. The more words, the larger the representation of documents, therefore it is important to constrain the words to only those believed to be predictive. 

We can build a vocabulary as a `Counter` from the `collections` library, which is a dictionary mapping of words and their count that allows us to easily update and query. Each document can be added to the counter (a new function called `add_doc_to_vocab()`) and we can step over all of the reviews in the negative directory and then the positive directory (a new function called `load_dir()`). 

In [19]:

import string
import re
from os import listdir
from collections import Counter 
from nltk.corpus import stopwords

# load doc into memory
def load_doc(filename):
# open the file as read only 
    file = open(filename, 'r')
# read all text
    text = file.read()
    # close the file 
    file.close()
    return text

# turn a doc into clean tokens
def clean_doc(doc):
# split into tokens by white space
    tokens = doc.split()
    # prepare regex for char filtering
    re_punc = re.compile('[%s]' % re.escape(string.punctuation)) 
    # remove punctuation from each word
    tokens = [re_punc.sub('', w) for w in tokens]
    # remove remaining tokens that are not alphabetic
    tokens = [word for word in tokens if word.isalpha()]
    # filter out stop words
    stop_words = set(stopwords.words('english'))
    tokens = [w for w in tokens if not w in stop_words]
    # filter out short tokens
    tokens = [word for word in tokens if len(word) > 1]
    return tokens

# load doc and add to vocab
def add_doc_to_vocab(filename, vocab): 
    # load doc
    doc = load_doc(filename)
    # clean doc
    tokens = clean_doc(doc)
      # update counts
    vocab.update(tokens)
    
    
# load all docs in a directory
def process_docs(directory, vocab):
# walk through all files in the folder 
    for filename in listdir(directory):
    # skip any reviews in the test set
        if filename.startswith('cv9'): 
            continue
        # create the full path of the file to open
        path = directory + '/' + filename 
        # add doc to vocab 
        add_doc_to_vocab(path, vocab)
        
        
# define vocab
vocab = Counter()
# add all docs to vocab
process_docs('../input/movie-review/movie_reviews/movie_reviews/pos', vocab) 
process_docs('../input/movie-review/movie_reviews/movie_reviews/neg', vocab) 
# print the size of the vocab 
print(len(vocab))
# print the top words in the vocab
print(vocab.most_common(100))

44276
[('film', 7983), ('one', 4946), ('movie', 4826), ('like', 3201), ('even', 2262), ('good', 2080), ('time', 2041), ('story', 1907), ('films', 1873), ('would', 1844), ('much', 1824), ('also', 1757), ('characters', 1735), ('get', 1724), ('character', 1703), ('two', 1643), ('first', 1588), ('see', 1557), ('way', 1515), ('well', 1511), ('make', 1418), ('really', 1407), ('little', 1351), ('life', 1334), ('plot', 1288), ('people', 1269), ('could', 1248), ('bad', 1248), ('scene', 1241), ('movies', 1238), ('never', 1201), ('best', 1179), ('new', 1140), ('scenes', 1135), ('man', 1131), ('many', 1130), ('doesnt', 1118), ('know', 1092), ('dont', 1086), ('hes', 1024), ('great', 1014), ('another', 992), ('action', 985), ('love', 977), ('us', 967), ('go', 952), ('director', 948), ('end', 946), ('something', 945), ('still', 936), ('seems', 930), ('back', 923), ('made', 912), ('theres', 901), ('work', 901), ('makes', 883), ('however', 874), ('years', 867), ('world', 864), ('every', 858), ('big', 8

Running the snippet above shows that we have a vocabulary of 44,276 words. We also can see a sample of the top 100 most used words in the movie reviews. The next step would be to develop a function to save the cleaned text into a file.

In [20]:
def save_list(lines, filename):
    # convert lines into blob text
    data = '\n'.join(lines)
    # open file
    file = open(filename, 'w')
    file.write(data)
    file.close

In [21]:
vocab = Counter()
# add all docs to vocab
process_docs('../input/movie-review/movie_reviews/movie_reviews/pos', vocab) 
process_docs('../input/movie-review/movie_reviews/movie_reviews/neg', vocab) 

min_occurance = 2
tokens = [k for k, c in vocab.items() if c >= min_occurance]
print(len(tokens))

save_list(tokens, 'vocab.txt')

25767


## Train CNN with Embedding Layer

In this section, we will explore the way of word embedding while training a convolutional neural network on the classification problem. 

A word embedding is a way of representing the text where each word in the vocabulary is represented by a real valued vector in a high dimensional space.

Words that have similar meaning would have similar presentation in the vector space.

The real valued vector representation for words can be learned while training the neural network. We can do this in the Keras deep learning library using the `Embedding layer`. 

The first step is to load the vocabulary. We will use it to filter out words from movie reviews that we are not interested in. We already did that when we save filtered tokens in `vocab.txt` file, we can load it using the `load_doc()`

In [22]:
vocab_filename = '../input/cleaned-token/vocab.txt'
vocab = load_doc(vocab_filename)
vocab = set(vocab.split())

Next, we need to load all of the training data movie reviews. For that we can adapt the process docs() from the previous section to load the documents, clean them, and return them as a list of strings, with one document per string. We want each document to be a string for easy encoding as a sequence of integers later. Cleaning the document involves splitting each review based on white space, removing punctuation, and then filtering out all tokens not in the vocabulary. 

In [23]:
# turn doc into a clean token
def clean_doc(doc, vocab):
    # split into tokens by white space
    tokens = doc.split()
    # prepare regex for char filtering
    re_punc = re.compile('[%s]' % re.escape(string.punctuation)) 
    # remove punctuation from each word
    tokens = [re_punc.sub('', w) for w in tokens]
    # filter out tokens not in vocab
    tokens = [w for w in tokens if w in vocab]
    tokens = ' '.join(tokens)
    return tokens

In [24]:
# load all docs in a directory
def process_docs(directory, vocab, is_train): 
    documents = list()
    # walk through all files in the folder
    for filename in listdir(directory):
        # skip any reviews in the test set
        if is_train and filename.startswith('cv9'): 
            continue
        if not is_train and not filename.startswith('cv9'): 
            continue
        # create the full path of the file to open
        path = directory + '/' + filename 
        # load the doc
        doc = load_doc(path)
        # clean doc
        tokens = clean_doc(doc, vocab)
        # add to list
        documents.append(tokens) 
    return documents

Combine `neg` and `pos` reviews into one train dataset. 

In [25]:
# load and clean a dataset
def load_clean_dataset(vocab, is_train):
    # load documents
    neg = process_docs('../input/movie-review/movie_reviews/movie_reviews/neg', vocab, is_train)
    pos = process_docs('../input/movie-review/movie_reviews/movie_reviews/pos', vocab, is_train)
    docs = neg + pos
    # prepare labels
    labels = [0 for _ in range(len(neg))] + [1 for _ in range(len(pos))]
    return docs, labels

The next step is to encode each document as a sequence of integers. The Keras Embedding layer requires integer inputs where each integer maps to a single token that has a specific real-valued vector representation within the embedding. These vectors are random at the beginning of training, but during training become meaningful to the network. We can encode the training documents as sequences of integers using the Tokenizer class in the Keras API.

In [26]:
from keras.preprocessing.text import Tokenizer
def create_tokenizer(lines):
    tokenizer = Tokenizer()
    tokenizer.fit_on_texts(lines)
    return tokenizer

In [27]:
print(Xtrain.shape, Xtest.shape)

(1800, 1317) (200, 1317)


Now that the mapping of words to integers has been prepared, we can use it to encode the reviews in the training dataset. We can do that by calling the texts to sequences() function on the Tokenizer. We also need to ensure that all documents have the same length. This is a requirement of Keras for efficient computation.

In [28]:
max_length = max([len(s.split()) for s in train_docs])
print('Maximum length: %d' % max_length)

Maximum length: 1317


In [29]:
import string
import re
from os import listdir
from numpy import array
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences 
from keras.utils.vis_utils import plot_model
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Flatten
from keras.layers import Embedding
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D


# load doc into memory
def load_doc(filename):
    # open the file as read only 
    file = open(filename, 'r')
    # read all text
    text = file.read()
    # close the file 
    file.close()
    return text


# turn a doc into clean tokens
def clean_doc(doc, vocab):
    # split into tokens by white space
    tokens = doc.split()
    # prepare regex for char filtering
    re_punc = re.compile('[%s]' % re.escape(string.punctuation)) 
    # remove punctuation from each word
    tokens = [re_punc.sub('', w) for w in tokens]
    # filter out tokens not in vocab
    tokens = [w for w in tokens if w in vocab]
    tokens = ' '.join(tokens)
    return tokens


# load all docs in a directory
def process_docs(directory, vocab, is_train): 
    documents = list()
    # walk through all files in the folder
    for filename in listdir(directory):
        # skip any reviews in the test set
        if is_train and filename.startswith('cv9'): 
            continue
        if not is_train and not filename.startswith('cv9'): 
            continue
            # create the full path of the file to open
        path = directory + '/' + filename # load the doc
        doc = load_doc(path)
        # clean doc
        tokens = clean_doc(doc, vocab)
        # add to list
        documents.append(tokens) 
    return documents


# load and clean a dataset
def load_clean_dataset(vocab, is_train):
    # load documents
    neg = process_docs('../input/movie-review/movie_reviews/movie_reviews/neg', vocab, is_train)
    pos = process_docs('../input/movie-review/movie_reviews/movie_reviews/pos', vocab, is_train)
    docs = neg + pos
    # prepare labels
    labels = array([0 for _ in range(len(neg))] + [1 for _ in range(len(pos))]) 
    return docs, labels


# fit a tokenizer
def create_tokenizer(lines): 
    tokenizer = Tokenizer() 
    tokenizer.fit_on_texts(lines) 
    return tokenizer


# integer encode and pad documents
def encode_docs(tokenizer, max_length, docs): 
    # integer encode
    encoded = tokenizer.texts_to_sequences(docs) 
    # pad sequences
    padded = pad_sequences(encoded, maxlen=max_length, padding='post') 
    return padded


# define the model
def define_model(vocab_size, max_length):
    model = Sequential()
    model.add(Embedding(vocab_size, 100, input_length=max_length)) 
    model.add(Conv1D(filters=32, kernel_size=8, activation='relu')) 
    model.add(MaxPooling1D(pool_size=2))
    model.add(Flatten())
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    # compile network
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy']) 
    # summarize defined model
    model.summary()
    plot_model(model, to_file='model.png', show_shapes=True)
    return model


# load the vocabulary
vocab_filename = '../input/cleaned-token/vocab.txt'
vocab = load_doc(vocab_filename)
vocab = set(vocab.split())
# load training data
train_docs, ytrain = load_clean_dataset(vocab, True) 
# create the tokenizer
tokenizer = create_tokenizer(train_docs)
# define vocabulary size
vocab_size = len(tokenizer.word_index) + 1
print('Vocabulary size: %d' % vocab_size)
# calculate the maximum sequence length
max_length = max([len(s.split()) for s in train_docs]) 
print('Maximum length: %d' % max_length)
# encode data
Xtrain = encode_docs(tokenizer, max_length, train_docs) 
# define model
model = define_model(vocab_size, max_length)
# fit network
model.fit(Xtrain, ytrain, epochs=10, verbose=2)
# save the model
model.save('model.h5')

Vocabulary size: 25768
Maximum length: 1317
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 1317, 100)         2576800   
_________________________________________________________________
conv1d_1 (Conv1D)            (None, 1310, 32)          25632     
_________________________________________________________________
max_pooling1d_1 (MaxPooling1 (None, 655, 32)           0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 20960)             0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                209610    
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 11        
Total params: 2,812,053
Trainable params: 2,812,053
Non-trainable params: 0


## Model evaluation
we will evaluate the trained model and use it to make predictions on new data. First, we can use the built-in evaluate() function to estimate the skill of the model on both the training and test dataset. This requires that we load and encode both the training and test datasets.

In [30]:
from keras.models import load_model
# load all reviews
train_docs, ytrain = load_clean_dataset(vocab, True)
test_docs, ytest = load_clean_dataset(vocab, False)
# create the tokenizer
tokenizer = create_tokenizer(train_docs)
# define vocabulary size
vocab_size = len(tokenizer.word_index) + 1 
print('Vocabulary size: %d' % vocab_size)
# calculate the maximum sequence length
max_length = max([len(s.split()) for s in train_docs]) 
print('Maximum length: %d' % max_length)
# encode data
Xtrain = encode_docs(tokenizer, max_length, train_docs) 
Xtest = encode_docs(tokenizer, max_length, test_docs)
# load the model
model = load_model('../input/generated-model/model.h5')
# evaluate model on training dataset
_, acc = model.evaluate(Xtrain, ytrain, verbose=0) 
print('Train Accuracy: %f' % (acc*100))
# evaluate model on test dataset
_, acc = model.evaluate(Xtest, ytest, verbose=0) 
print('Test Accuracy: %f' % (acc*100))

Vocabulary size: 25768
Maximum length: 1317
Train Accuracy: 86.166668
Test Accuracy: 70.499998


New data must then be prepared using the same text encoding and encoding schemes as was used on the training dataset. Once prepared, a prediction can be made by calling the predict() function on the model. The function below named predict sentiment() will encode and pad a given movie review text and return a prediction in terms of both the percentage and a label.


In [31]:
# classify a review as negative or positive
def predict_sentiment(review, vocab, tokenizer, max_length, model): 
    # clean review
    line = clean_doc(review, vocab)
    # encode and pad review
    padded = encode_docs(tokenizer, max_length, [line])
      # predict sentiment
    yhat = model.predict(padded, verbose=0)
      # retrieve predicted percentage and label
    percent_pos = yhat[0,0]
    if round(percent_pos) == 0:
        return (1-percent_pos), 'NEGATIVE' 
    return percent_pos, 'POSITIVE'

In [32]:
text = 'Everyone will enjoy this film. I love it, recommended!'
percent, sentiment = predict_sentiment(text, vocab, tokenizer, max_length, model) 
print('Review: [%s]\nSentiment: %s (%.3f%%)' % (text, sentiment, percent*100))
# test negative text
text = 'This is a bad movie. Do not watch it. It sucks.'
percent, sentiment = predict_sentiment(text, vocab, tokenizer, max_length, model) 
print('Review: [%s]\nSentiment: %s (%.3f%%)' % (text, sentiment, percent*100))

Review: [Everyone will enjoy this film. I love it, recommended!]
Sentiment: NEGATIVE (50.019%)
Review: [This is a bad movie. Do not watch it. It sucks.]
Sentiment: NEGATIVE (51.373%)


Running the example first prints the skill of the model on the training and test dataset. We can see that the model achieves 86.16% accuracy on the training dataset and 70.4% on the test dataset, not a bad score.

Next, we can see that the model makes the correct prediction on two contrived movie reviews. We can see that the percentage or confidence of the prediction is close to 50% for both, this may be because the two contrived reviews are very short and the model is expecting sequences of 1,000 or more words